File: kotlin/exi-iso20-common/src/main/kotlin/cloud/charging/v2g/iso20/common/V2GSignature.kt
package cloud.charging.v2g.iso20.common

import org.bouncycastle.crypto.params.Ed448PrivateKeyParameters
import org.bouncycastle.crypto.params.Ed448PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed448Signer
import java.security.MessageDigest
import java.security.PrivateKey
import java.security.PublicKey
import java.security.Signature

/**
 * XMLDSig signing and verification for ISO 15118-20 CommonMessages.
 *
 * The same two-level EXI-fragment digest scheme as -2's `V2GSignature` — see there for the general
 * shape — but with -20's stronger suite: SHA-512 digests, and either of the two asymmetric
 * algorithms the -20 signature suite allows:
 *
 *  * **ECDSA over NIST P-521** (secp521r1), raw `r‖s` SignatureValue, 132 bytes (66 + 66,
 *    IEEE P1363) — [sign] / [verify];
 *  * **Ed448** (RFC 8032), raw 114-byte SignatureValue — [signEd448] / [verifyEd448].
 *
 * Ed448 is why this module depends on BouncyCastle: the JDK has no Ed448 at all — not a missing
 * provider, the algorithm simply is not there — while it does have Ed25519. The C# side reaches
 * for the same library for the same reason.
 *
 * Ed448 is a *pure* EdDSA scheme, so it signs the SignedInfo fragment octets directly: its internal
 * SHAKE256 takes the place of an external pre-hash, and there is no separate SHA-512 step the way
 * [sign] has one. The SHA-512 in [digest] is a different thing — it hashes the *signed elements*
 * into their References, and applies whichever signature algorithm is chosen.
 */
object V2GSignature {

    /** EXI canonicalization — the only C14N ISO 15118-20 uses. */
    const val CANONICALIZATION_EXI = "http://www.w3.org/TR/canonical-exi/"

    /** ECDSA-SHA512 signature method. */
    const val ECDSA_SHA512 = "http://www.w3.org/2001/04/xmldsig-more#ecdsa-sha512"

    /** Ed448 (pure EdDSA, RFC 8032) signature method — RFC 9231 §2.3.12. */
    const val EDDSA_ED448 = "http://www.w3.org/2021/04/xmldsig-more#eddsa-ed448"

    /** SHA-512 digest method. */
    const val SHA512 = "http://www.w3.org/2001/04/xmlenc#sha512"

    /**
     * JCA name for ECDSA-SHA512 producing the raw `r‖s` form ISO 15118 requires, rather than the
     * ASN.1/DER the plain name yields. Available from Java 9 in the SunEC provider.
     */
    private const val JCA_ECDSA_P1363 = "SHA512withECDSAinP1363Format"

    /** SHA-512 of an element's EXI fragment — the value that goes into its `DigestValue`. */
    fun digest(fragmentBytes: ByteArray): ByteArray =
        MessageDigest.getInstance("SHA-512").digest(fragmentBytes)

    /**
     * A single-reference [SignedInfoType] over one already-computed element digest, with the fixed
     * EXI-C14N / SHA-512 algorithm URIs and the given signature method ([ECDSA_SHA512] by default,
     * or [EDDSA_ED448]). The reference URI is `"#" + referenceId` — the `Id` attribute of the
     * signed element.
     *
     * [includeExiTransform] adds the schema-optional `Transforms` list holding the single EXI-C14N
     * transform; some peers (Josev's pydantic models) treat it as mandatory and reject a Reference
     * without it.
     */
    fun buildSignedInfo(
        referenceId: String,
        digest: ByteArray,
        signatureMethodAlgorithm: String = ECDSA_SHA512,
        includeExiTransform: Boolean = false,
    ): SignedInfoType =
        buildSignedInfo(listOf(referenceId to digest), signatureMethodAlgorithm, includeExiTransform)

    /** A multi-reference [SignedInfoType] — one Reference per signed element, one signature. */
    fun buildSignedInfo(
        references: List<Pair<String, ByteArray>>,
        signatureMethodAlgorithm: String = ECDSA_SHA512,
        includeExiTransform: Boolean = false,
    ): SignedInfoType =
        SignedInfoType(
            id = null,
            canonicalizationMethod = CanonicalizationMethodType(algorithm = CANONICALIZATION_EXI, aNY = null),
            signatureMethod = SignatureMethodType(
                algorithm = signatureMethodAlgorithm, hMACOutputLength = null, aNY = null,
            ),
            reference = references.map { (referenceId, digest) ->
                ReferenceType(
                    id = null,
                    type = null,
                    uRI = "#$referenceId",
                    transforms = if (includeExiTransform)
                        TransformsType(listOf(TransformType(CANONICALIZATION_EXI, xPath = null, aNY = null)))
                    else null,
                    digestMethod = DigestMethodType(algorithm = SHA512, aNY = null),
                    digestValue = digest,
                )
            },
        )

    /** The header [SignatureType] from a signed `SignedInfo` and its raw SignatureValue. */
    fun buildSignature(signedInfo: SignedInfoType, signatureValue: ByteArray): SignatureType =
        SignatureType(
            id = null,
            signedInfo = signedInfo,
            signatureValue = SignatureValueType(id = null, value = signatureValue),
            keyInfo = null,
            `object` = null,
        )

    /** The exact octets that are signed (or verified): the `SignedInfo` as an EXI fragment. */
    fun signedInfoFragment(signedInfo: SignedInfoType): ByteArray =
        CommonMessagesCodec.encodeFragment_SignedInfo(signedInfo)

    /**
     * Signs a [SignedInfoType]: SHA-512 over its EXI fragment, ECDSA-P521, returning the raw `r‖s`
     * (132-byte) SignatureValue.
     */
    fun sign(signedInfo: SignedInfoType, privateKey: PrivateKey): ByteArray =
        Signature.getInstance(JCA_ECDSA_P1363).run {
            initSign(privateKey)
            update(signedInfoFragment(signedInfo))
            sign()
        }

    /**
     * Verifies a raw `r‖s` SignatureValue against a [SignedInfoType] and public key. This checks
     * only the signature over the SignedInfo fragment — confirming that each Reference digest
     * matches its signed element is [verifyReference]'s job, and skipping it would accept a
     * correctly signed SignedInfo attached to entirely different content.
     */
    fun verify(signedInfo: SignedInfoType, signatureValue: ByteArray, publicKey: PublicKey): Boolean =
        try {
            Signature.getInstance(JCA_ECDSA_P1363).run {
                initVerify(publicKey)
                update(signedInfoFragment(signedInfo))
                verify(signatureValue)
            }
        } catch (_: java.security.SignatureException) {
            // A malformed r‖s is a failed verification, not a crash.
            false
        }

    /**
     * Signs a [SignedInfoType] with Ed448 (RFC 8032), returning the raw 114-byte SignatureValue.
     * Pure EdDSA: the fragment octets are signed directly, with no external pre-hash.
     */
    fun signEd448(signedInfo: SignedInfoType, privateKey: Ed448PrivateKeyParameters): ByteArray {
        val message = signedInfoFragment(signedInfo)
        // The empty byte array is Ed448's context string (RFC 8032 §5.2), which ISO 15118-20 leaves
        // empty. A non-empty context would produce signatures a conforming peer cannot verify.
        return Ed448Signer(ByteArray(0)).apply {
            init(true, privateKey)
            update(message, 0, message.size)
        }.generateSignature()
    }

    /**
     * Verifies a raw 114-byte Ed448 SignatureValue against a [SignedInfoType] and public key. As
     * with [verify], this covers the SignedInfo only — see [verifyReference] for the other half.
     */
    fun verifyEd448(
        signedInfo: SignedInfoType,
        signatureValue: ByteArray,
        publicKey: Ed448PublicKeyParameters,
    ): Boolean {
        val message = signedInfoFragment(signedInfo)
        return Ed448Signer(ByteArray(0)).apply {
            init(false, publicKey)
            update(message, 0, message.size)
        }.verifySignature(signatureValue)
    }

    /**
     * Confirms that a Reference's `DigestValue` equals the SHA-512 of the given signed-element
     * fragment — the second half of verification. [MessageDigest.isEqual] is constant-time.
     */
    fun verifyReference(reference: ReferenceType, signedElementFragment: ByteArray): Boolean =
        MessageDigest.isEqual(reference.digestValue, digest(signedElementFragment))
}

File: swift/Sources/ExiIso20DC/V2GSignature.swift
#if canImport(CryptoKit)
import CryptoKit
import Foundation

/// The ISO 15118-20 DC half of the -20 signature suite: SHA-512 over a signed element's EXI
/// **fragment** goes into a `SignedInfo` Reference, and the `SignedInfo` fragment is itself signed.
///
/// ## Why every set carries its own copy
///
/// This helper is repeated across the -20 sets, and that is not laziness. Each set embeds its own
/// copy of the XMLDSig schema, and the fragment grammar's element selector is sized by the whole
/// set — so the *same* `SignedInfo` lands on a different event code in each. Here it is **129 over 8 bits**.
/// Different octets, therefore different signed bytes. Borrowing another set's helper would sign
/// something no peer asked for, and the signature would verify locally and nowhere else. Note the
/// DER sets move too, even though their *messages* are AC's.
///
/// ## What is implemented, and what is not
///
/// ISO 15118-20 allows two signature algorithms. Only the ECDSA one is here:
///
/// * **ECDSA over secp521r1 with SHA-512** — CryptoKit has P-521 natively, and exposes the raw
///   `r‖s` pair the wire wants as `rawRepresentation`. 66 + 66 bytes.
/// * **Ed448** (RFC 8032, 114 bytes) — **not available.** CryptoKit does not have the curve at
///   all; this is not a matter of an unregistered provider as it would be on the JVM, where
///   BouncyCastle supplies it. Implementing it needs a bundled crypto library (BC-Swift, OpenSSL
///   or wolfSSL), which is a dependency and binary-size decision for the app rather than anything
///   this layer can settle. See `docs/CONCEPT.md` §3.3.
///
/// A peer that selects Ed448 therefore cannot be answered by this build, and
/// ``sign(_:with:)`` is the only signing path there is — rather than an Ed448 stub that would
/// look implemented.
public enum V2GSignature {

    /// The wire length of a -20 ECDSA signature: r and s, 66 bytes each.
    public static let ecdsaSignatureLength = 132

    /// The wire length an Ed448 signature would have, kept for the check below.
    public static let ed448SignatureLength = 114

    /// SHA-512 over an element's EXI fragment — the digest a `SignedInfo` Reference carries.
    ///
    /// The input must be a *fragment*, not a message: a document-wrapped encoding of the same
    /// content digests differently and would be rejected by any conforming peer.
    public static func digest(ofFragment fragment: [UInt8]) -> [UInt8] {
        Array(SHA512.hash(data: fragment))
    }

    /// Signs a `SignedInfo` over its own EXI fragment with ECDSA-P521, returning the raw `r‖s`
    /// pair.
    public static func sign(_ signedInfo: SignedInfoType,
                            with key: P521.Signing.PrivateKey) throws -> [UInt8] {
        let fragment = DCCodec.encodeFragment_SignedInfo(signedInfo)
        return Array(try key.signature(for: Data(fragment)).rawRepresentation)
    }

    /// Verifies a raw `r‖s` ECDSA-P521 signature over a `SignedInfo`'s own EXI fragment.
    ///
    /// A signature of any other length is refused before CryptoKit sees it. That covers the DER
    /// form, which a JCA- or OpenSSL-minded implementation produces by default and which verifies
    /// against itself perfectly — and an Ed448 signature, which this build cannot check at all and
    /// must not silently report as invalid *cryptography* when the real answer is "unsupported
    /// algorithm".
    public static func verify(_ signedInfo: SignedInfoType,
                              signature: [UInt8],
                              with key: P521.Signing.PublicKey) throws -> Bool {
        if signature.count == ed448SignatureLength {
            throw V2GSignatureError.ed448NotAvailable
        }
        guard signature.count == ecdsaSignatureLength,
              let parsed = try? P521.Signing.ECDSASignature(rawRepresentation: Data(signature))
        else { return false }

        let fragment = DCCodec.encodeFragment_SignedInfo(signedInfo)
        return key.isValidSignature(parsed, for: Data(fragment))
    }
}

public enum V2GSignatureError: Error, Equatable {
    /// The signature is Ed448-shaped. The algorithm is part of the -20 suite but CryptoKit has no
    /// Ed448, so this build can neither produce nor check one — reported as a distinct condition so
    /// a caller can tell "unsupported" from "invalid".
    case ed448NotAvailable
}
#endif
